### PR TITLE
Add the key_id_header module and start on functions for dealing with it.

### PR DESCRIPTION
**Pull request by coltfred**
Adds create, encode and decode functions for dealing with key_id_headers. My plan is to leave this PR open until we implement the deterministic, vector and ironcore document versions in the SDK to ensure compatibility. Opening so we can get early feedback.

### CODE REVIEW

File: src/key_id_header.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
+use bytes::Bytes;
+use itertools::Itertools;
+use protobuf::Message;
+
+use crate::{vector_encryption_metadata::VectorEncryptionMetadata, Error};
+
+// This file is for functions which are working with our key id header value.
+// This value has the following structure:
+// 4 Byte id. This value is a u32 encoded in big endian format.
+// 1 Byte where the first 4 bits are used for which type of edek the id points to (Standalone, Saas Shield, DCP).
```
Comment (giarc3) [inline src/key_id_header.rs:11]: This comment says DCP is included but it's not below

File: src/key_id_header.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        }
+    }
+    pub fn create_header(&self, key_id: KeyId) -> Bytes {
+        let mut vec = Vec::with_capacity(6);
+        for b in u32::to_be_bytes(key_id.0) {
+            vec.push(b);
+        }
+        vec.push(self.to_numeric_value());
+        vec.push(0u8);
+        vec.into()
```
Comment (giarc3) [inline src/key_id_header.rs]: Thoughts on this? I think they're pretty similar, but I tend towards chaining
```suggestion
        let iter = u32::to_be_bytes(key_id.0)
            .into_iter()
            .chain([self.to_numeric_value(), 0u8]);
        Bytes::from_iter(iter)
```

File: src/key_id_header.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    pub(crate) fn parse_from_bytes(b: Bytes) -> Result<EdekType> {
+        b.first()
+            .ok_or_else(|| Error::EdekTypeError("EdekType couldn't be determined.".to_string()))
```
Comment (giarc3) [inline src/key_id_header.rs]: Do we want to be any more specific in this error? Could say here that the bytes are empty

File: src/key_id_header.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            .and_then(Self::from_numeric_value)
+    }
+}
+
+// TODO: Payload type not implemented here yet.
+pub fn create_vector_metadata(
```
Comment (giarc3) [inline src/key_id_header.rs:165]: The places you use this so far in the cloaked-ai PR call this function then pass the results straight to the `encode_vector_metadata` function. Should the `encode_vector_metadata` function actually just do both things (with a different name)? Will we need both functions exposed in the future?

File: src/key_id_header.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            Ok(EdekType::SaasShield)
+        } else if candidate == &STANDALONE_EDEK_TYPE_NUM {
+            Ok(EdekType::Standalone)
+        } else {
+            Err(Error::EdekTypeError(
+                "Byte {candidate} isn't a valid edek type.".to_string(),
+            ))
+        }
+    }
+    pub fn create_header(&self, key_id: KeyId) -> Bytes {
```
Comment (giarc3) [inline src/key_id_header.rs]: Needs a doc comment for public function

File: src/key_id_header.rs (hunk 4 above)
Comment (giarc3) [reply src/key_id_header.rs:165]: Also needs a doc comment

File: src/key_id_header.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        };
+        // What's left in header is 2 bytes for the padding
+        let edek_type = EdekType::parse_from_bytes(padding_bytes)?;
+        Ok((id, edek_type, rest))
+    } else {
+        Err(Error::KeyIdHeaderTooShort(value_len))
+    }
+}
+
+pub fn get_prefix_bytes_for_search(key_id: KeyId, edek_type: EdekType) -> Bytes {
```
Comment (giarc3) [inline src/key_id_header.rs]: Needs a doc comment

File: src/key_id_header.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+use protobuf::Message;
+
+use crate::{vector_encryption_metadata::VectorEncryptionMetadata, Error};
+
+// This file is for functions which are working with our key id header value.
+// This value has the following structure:
+// 4 Byte id. This value is a u32 encoded in big endian format.
+// 1 Byte where the first 4 bits are used for which type of edek the id points to (Standalone, Saas Shield, DCP).
+//   The next 4 bits are to denote which type of data follows it (vector metadata, IronCore Edoc, deterministic ciphertext)
+
```
Comment (giarc3) [inline src/key_id_header.rs:14]: Followed by one byte of `0`?

**Review by giarc3 — CHANGES_REQUESTED** [review]

File: src/key_id_header.rs (hunk 4 above)
Comment (coltfred) [reply src/key_id_header.rs:165]: I don't know if we'll ever need them separate or not, but they are for now.

Comment (giarc3) [reply src/key_id_header.rs:165]: Okay. In my head it still makes cleanliness sense for one to just call the other (even if we leave both exposed), but it's fine

**Review by giarc3 — APPROVED** [review]

**Pull request merged**